Source/Core/VideoCommon/BPFunctions.h

```diff
@@ -7,16 +7,154 @@
 
 #pragma once
 
-#include "Common/MathUtil.h"
+#include <utility>
+#include <vector>
 
-struct BPCmd;
+#include "Common/MathUtil.h"
+#include "VideoCommon/BPMemory.h"
+struct XFMemory;
 
 namespace BPFunctions
 {
+struct ScissorRange
+{
+  constexpr ScissorRange() = default;
+  constexpr ScissorRange(int offset, int start, int end) : offset(offset), start(start), end(end) {}
+  int offset = 0;
+  int start = 0;
+  int end = 0;
+};
+
+struct ScissorRect
+{
+  constexpr ScissorRect(ScissorRange x_range, ScissorRange y_range)
+      :  // Rectangle ctor takes x0, y0, x1, y1.
+        rect(x_range.start, y_range.start, x_range.end, y_range.end), x_off(x_range.offset),
+        y_off(y_range.offset)
+  {
+  }
+
+  MathUtil::Rectangle<int> rect;
+  int x_off;
+  int y_off;
+
+  int GetArea() const;
+};
+
+// Although the GameCube/Wii have only one scissor configuration and only one viewport
+// configuration, some values can result in multiple parts of the screen being updated.
+// This can happen if the scissor offset combined with the bottom or right coordinate ends up
+// exceeding 1024; then, both sides of the screen will be drawn to, while the middle is not.
+// Major Minor's Majestic March causes this to happen during loading screens and other scrolling
+// effects, though it draws on top of one of them.
+// This can also happen if the scissor rectangle is particularly large, but this will usually
+// involve drawing content outside of the viewport, which Dolphin does not currently handle.
+//
+// The hardware backends can currently only use one viewport and scissor rectangle, so we need to
```

Comment:
Could we disable scissoring on the host altogether and instead implement it manually in the pixel shaders instead to fully support this case?

Reply:
I don't think so.  Masking two different sides of the screen could be done using a pixel shader pretty easily, but I don't think moving things to different sides is possible in just a pixel shader.

Most graphics APIs do support multiple viewports and scissors internally, though, so I think it might be possible to implement it with that.  I just haven't looked into the details.

I added some additional images with the software renderer where the loading text is removed which might help visualize it: Frame 204, [ImGui](https://user-images.githubusercontent.com/8334194/143727516-8601edbe-7d7c-42e6-867a-19bc1ab1f055.png) and [SW](https://user-images.githubusercontent.com/8334194/145092021-4e9dcbac-2787-44de-a6f7-d55bb0fb6b4d.png); Frame 205, [ImGui](https://user-images.githubusercontent.com/8334194/143727517-4f08454b-7457-4a43-9996-f5646c4d6469.png) and [SW](https://user-images.githubusercontent.com/8334194/145092093-506dd5a7-bab1-4f7d-a9b5-067301195569.png).

Reply:
Hmm not sure I'm following. Do I understand you correctly there are two components that to such scissor "overflows"?
1. Pixels on the very left and on the very right are rendered, leaving a big gap in the middle
2. The left subarea is not a true scissor rect, but instead content that ends up on the left subarea is rendered *as if* it had been rendered to the right of the right subarea, i.e. that content wraps around

Reading the comment it was easy to see 1 being the case, but 2 comes at a surprise. Granted, I may have glanced over the comment too quickly. If I didn't just miss it, I think it's a detail worth mentioning though.

Reply:
I think I just haven't explained things very well (I didn't do a writeup beforehand, other than my messy notes); it definitely could be clearer.

Both 1 and 2 are correct.  But I think there might be a clearer explanation.

The scissor rectangle itself _doesn't_ overflow.  You can give it any set of values `x0`, `y0`, `x1`, and `y1` all in the range `[0, 2047]` and things will render correctly: if `x0 > x1` or `y0 > y1`, nothing shows up; otherwise, pixels between those ranges will be rasterized.  This is after the viewport has been applied, so normally you want the viewport and scissor to match (and both the scissor and viewport have 342 added to their origins by GX functions).  (The viewport is also based on floats, and uses a center+radius format, which is a bit confusing, but that can be ignored here).  For instance, a 640 by 480 viewport and scissor would have the scissor coordinates 342, 342, 981, 821, and a viewport for 342, 342, 982, 822.

You can still increase the size of the scissor, and things behave normally for any size (even 0, 0, 2047, 2047), except that you'll start getting stuff that's outside of the viewport and instead in the clipper guardband.  I suspect that if you change the viewport size to something larger, everything will behave normally, but in my testing I always used the same viewport.

The scissor offset is then used to choose where rasterized pixels end up in the EFB.  Usually, both `x_off` and `y_off` set to 171, which becomes an offset of 342 when multiplied by 2.  So the pixels that past the scissor test (those with x between 342 and 981 and y between 342 and 821) are offset by 342 into the EFB, ending up with x between 0 and 639 and y between 0 and 479.

The scissor offset step is the thing that can overflow.  Textures on the GameCube/Wii are at max 1024 by 1024, and I guess operations on the EFB wrap around at that point.  Suppose `x_off` was 427, which becomes an offset of 854 = 342 + 512.  Then, a scissor with `x0` of 342 and `x1` of 981 ends up at -512 through 127.  -512 becomes +512, and 512 is less than the EFB's width of 640, so the pixels at the left of the viewport (x=0 without the offset of 342) end up at x=512, and pixels at the right of the viewport (x=639 without the offset of 342) end up at x=127; x=512-639 and x=0-127 end up being drawn to and x=128-511 don't get drawn.

You _can_ also get overflowing with a default `x_off` of 342 if the scissor has `x1` - `x0` > 1024; for instance, in a scissor with `x0` of 342 and `x1` of 2005 (639+342+1024), pixels with x between 342 and 2005 all pass the scissor test and then both x between 342-981 and x between 1366-2005 map to 0-639 on screen.  With the default viewport, one of these is far out of bounds and will usually be clipped (always clipped in Dolphin, since it seems like we clip precisely to the viewport), but with clipping disabled (or presumably a large enough viewport) you can get primitives that draw over themselves, producing [weird results](https://i.imgur.com/6xwybM2.png).

All of this is also made more confusing by the fields for `x0`, `y0`, `x1`, and `y1` being large enough to store `[0, 4095]` despite actually only going for `[0, 2047]` and `x_off`/`y_off` being large enough to store `[0, 2046]` (in jumps of 2). but instead only covering `[0, 1022]` (still in jumps of 2).

Hopefully this is enough to at least understand why things are happening.  I'm not sure how to condense this down to a simpler, more useful explanation.

Reply:
Heh, what a mess. Not sure I can help simplify this any further, but thanks for the explanation :)